source/common/quic/active_quic_listener.h

```diff
@@ -20,6 +20,18 @@
 namespace Envoy {
 namespace Quic {
 
+class HotRestartPacketForwardingOptions : virtual public Network::ExtraShutdownListenerOptions {
```

Comment:
If you are not doing to distinguish QUIC listener from TCP ones, how would you know which drainListener() call should pass down this object instead of the place holder? Would it make more sense to pass down more general information, i.e. child process UNIX domain socket address, in ExtraShutdownListenerOptions for all types of listeners instead of passing down a QUIC specific one?

Reply:
This is what the dynamic typing is for - you can pass in an object containing *all* the special-case information for *all* types of listener (which right now is just the one), and the listeners can pick out the information they care about and ignore any information they don't care about. So the answer is the drainListener call that gets started by the hot restart classes can pass the same set of options down to all the resulting shutdownListeners calls.

Passing a domain socket in would mean the quic stuff would need to know about the whole hot restart RPC protocol, which would also make it super unwieldy to unit test.

Trying to pick out listeners by type in advance would be a large unwieldy change because the listener manager barely allows access to individual listeners at all, and would be putting a bunch of decision logic in a place where it doesn't really need to be.

The lambda part acts as an abstract interface that makes it easy to mock.

Reply:
> Passing a domain socket in would mean the quic stuff would need to know about the whole hot restart RPC protocol, which would also make it super unwieldy to unit test.
> 
> Trying to pick out listeners by type in advance would be a large unwieldy change because the listener manager barely allows access to individual listeners at all, and would be putting a bunch of decision logic in a place where it doesn't really need to be.

In light of these 2 statements, how do you plan to call shutdownListener() with the right ExtraShutdownListenerOptions implementation?

Reply:
Construct and pass in the options [here](https://sourcegraph.com/github.com/envoyproxy/envoy/-/blob/source/server/hot_restarting_parent.cc?L187). The options parameter needs to be expanded out another few layers, but not yet in this PR. Trying to keep it in small pieces.

Reply:
Okay, my suggestion is instead of instantiate a QUIC specific object there, removing the abstraction and adding an interface in ExtraShutdownListenerOptions and name it getUdpPacketForwardingFunction(). I assume as long as the underlying UDP listener want to forward packets during draining, either QUIC or UDP proxy, they should use the same way.

Reply:
The problem with getting rid of the abstraction is it's inflexible for any future custom behaviors we might want to support. If it's just an interface and we override it for this use-case, and then we have a different use-case later that wants to pass different things in from a different caller for a different listener, when we add another function to Options, the existing implementation now needs to override that function too, and doesn't have anything to override it with, and the new caller needs to provide an override for getUdpPacketForwardingFunction that it doesn't care about, and the normal caller passing nullptr means now we have to check if the options passed was nullptr *or* the function-getter returns nullptr, where with the dynamic_cast the existence of the interface implies the existence of the function.

OR you end up adding another parameter to all the shutdown functions for the new purpose of interface, which is even messier.

Compromise change, I've made it HotRestartUdpPacketForwarding, moved it out of the quic context, and made the interface part of the class so there's no lambdas involved now.